File: src/wasm-exporter.js
import { MEMORY, SLOTS, ZIG, SOURCE, STRUCTURE } from './symbol.js';
import {
  StructureType,
  beginStructure,
  attachMember,
  attachMethod,
  attachTemplate,
  finalizeStructure,
  getStructureFeature,
} from './structure.js';
import { MemberType, getMemberFeature } from './member.js';
import { decamelizeErrorName } from './error.js';
import { getMemoryCopier } from './memory.js';

export async function linkWASMBinary(binaryPromise, params = {}) {
  const {
    resolve,
    reject,
    ...linkParams
  } = params;
  try {
    const wasmBinary = await binaryPromise;
    const result = await runWASMBinary(wasmBinary, linkParams);
    resolve(result);
  } catch (err) {
    reject(err);
  }
}

export async function runWASMBinary(wasmBinary, options = {}) {
  const {
    omitFunctions,
    variables,
    methodRunner,
  } = options;
  let nextValueIndex = 1;
  let valueTable = { 0: null };
  const valueIndices = new WeakMap();
  let nextStringIndex = 1;
  const stringTable = { 0: null };
  const stringIndices = {};
  const decoder = new TextDecoder();
  const callContexts = {};
  const globalSlots = {};
  let nextLinkageIndex = 0;
  let nextMethodIndex = 0;
  const structures = [];
  const imports = {
    _startCall,
    _endCall,
    _allocMemory,
    _freeMemory,
    _getMemory,
    _getMemoryOffset,
    _getMemoryLength,
    _wrapMemory,
    _createString,
    _createObject,
    _setObjectPropertyString,
    _setObjectPropertyInteger,
    _setObjectPropertyBoolean,
    _setObjectPropertyObject,
    _getPointerStatus,
    _setPointerStatus,
    _readGlobalSlot,
    _writeGlobalSlot,
    _readObjectSlot,
    _writeObjectSlot,
    _beginStructure,
    _attachMember,
    _attachMethod,
    _attachTemplate,
    _finalizeStructure,
    _createDataView,
    _createTemplate,
    _createArray,
    _appendArray,
    _logValues,
  };
  const { instance } = await WebAssembly.instantiate(wasmBinary, { env: imports });
  const { memory: wasmMemory, init, run, get, alloc, free } = instance.exports;
  if (process.env.NODE_ZIG_TARGET === 'WASM-STAGE1') {
    init();
    // call factory function
    const argStructIndex = addObject({ [SLOTS]: {} });
    const errorIndex = run(argStructIndex, 0);
    if (errorIndex !== 0) {
      throwError(errorIndex);
    }
    return structures;
  } else if (process.env.NODE_ZIG_TARGET === 'WASM-STAGE2') {
    init();
    // link variables
    for (const [ index, target ] of variables.entries()) {
      const address = get(index);
      const temp = target[MEMORY];
      const len = temp.byteLength;
      const wasm = new DataView(wasmMemory.buffer, address, len);
      wasm[SOURCE] = wasmMemory;
      const copy = getMemoryCopier(len);
      // copy changes made to temp buffer into WASM memory
      copy(wasm, temp);
      Object.defineProperty(target, MEMORY, { value: wasm });
    }
    // link methods
    methodRunner[0] = function(thunkIndex, argStruct) {
      const argIndex = addObject(argStruct);
      const errorIndex = run(argIndex, thunkIndex);
      if (errorIndex !== 0) {
        throwError(errorIndex);
      }
    };
  } else {
    throw new Error(`The environment variable NODE_ZIG_TARGET must be "WASM-STAGE1" or "WASM-STAGE2"`);
  }

  function addString(address, len) {
    const ta = new Uint8Array(wasmMemory.buffer, address, len);
    const s = decoder.decode(ta);
    let index = stringIndices[s];
    if (index === undefined) {
      index = stringIndices[s] = nextStringIndex++;
      stringTable[index] = s;
    }
    return index;
  }

  function addObject(object) {
    const index = nextValueIndex++;
    valueTable[index] = object;
    valueIndices.set(object, index);
    return index;
  }

  function throwError(errorIndex) {
    const errorName = stringTable[errorIndex];
    const errorMsg = decamelizeErrorName(errorName);
    throw new Error(errorMsg);
  }

  function _startCall(ctxAddr) {
    callContexts[ctxAddr] = { bufferMap: new Map() };
  }

  function _endCall(ctxAddr) {
    // move data from WASM memory into buffers
    const ctx = callContexts[ctxAddr];
    for (const [ buffer, { address, len, dv, copy } ] of ctx.bufferMap) {
      const src = new DataView(wasmMemory.buffer, address, len);
      copy(dv, src);
    }
    delete callContexts[ctxAddr];
    if (Object.keys(callContexts) === 0) {
      // clear the value table
      nextValueIndex = 1;
      valueTable = { 0: null };
    }
  }

  function _allocMemory(ctxAddr, len) {
    const address = alloc(ctxAddr, len);
    const ctx = callContexts[ctxAddr];
    const buffer = new ArrayBuffer(len);
    const dv = new DataView(buffer);
    const copy = getMemoryCopier(len);
    ctx.bufferMap.set(buffer, { address, len, dv, copy });
    return address;
  }

  function _freeMemory(ctxAddr, address, len) {
    const ctx = callContexts[ctxAddr];
    for (const [ buffer, { address: matching } ] of bufferMap) {
      if (address === matching) {
        bufferMap.delete(buffer);
        free(ctxAddr, address, len);
      }
    }
  }

  function _getMemory(ctxAddr, objectIndex) {
    const object = valueTable[objectIndex];
    let dv = object[MEMORY];
    if (!dv) {
      return 0;
    }
    const ctx = callContexts[ctxAddr];
    let memory = ctx.bufferMap.get(dv.buffer);
    if (!memory) {
      const len = dv.buffer.byteLength;
      const address = alloc(ctxAddr, len);
      const dest = new DataView(wasmMemory.buffer, address, len);
      // create new dataview if the one given only covers a portion of it
      const src = (dv.byteLength === len) ? dv : new DataView(dv.buffer);
      const copy = getMemoryCopier(len);
      copy(dest, src);
      memory = { address, len, dv: src, copy };
      ctx.bufferMap.set(dv.buffer, memory);
    }
    return addObject({
      address: memory.address + dv.byteOffset,
      len: dv.byteLength
    });
  }

  function _getMemoryOffset(objectIndex) {
    const object = valueTable[objectIndex];
    return object.address;
  }

  function _getMemoryLength(objectIndex) {
    const object = valueTable[objectIndex];
    return object.len;
  }

  function _wrapMemory(structureIndex, viewIndex) {
    const structure = valueTable[structureIndex];
    const dv = valueTable[viewIndex];
    let object;
    if (process.env.NODE_ZIG_TARGET === 'WASM-STAGE1') {
      object = {
        [STRUCTURE]: structure,
        [MEMORY]: dv,
        [SLOTS]: {},
      };
      if (structure.type === StructureType.Pointer) {
        object[ZIG] = true;
      }
    } else {
      const { constructor } = structure;
      object = constructor.call(ZIG, dv);
    }
    return addObject(object);
  }

  function _createString(address, len) {
    return addString(address, len);
  }

  function _createObject() {
    return addObject({});
  }

  function _setObjectPropertyString(containerIndex, keyIndex, valueIndex) {
    const container = valueTable[containerIndex];
    const key = stringTable[keyIndex];
    const value = stringTable[valueIndex];
    container[key] = value;
  }

  function _setObjectPropertyInteger(containerIndex, keyIndex, value) {
    const container = valueTable[containerIndex];
    const key = stringTable[keyIndex];
    container[key] = value;
  }

  function _setObjectPropertyBoolean(containerIndex, keyIndex, value) {
    const container = valueTable[containerIndex];
    const key = stringTable[keyIndex];
    container[key] = !!value;
  }

  function _setObjectPropertyObject(containerIndex, keyIndex, valueIndex) {
    const container = valueTable[containerIndex];
    const key = stringTable[keyIndex];
    container[key] = valueTable[valueIndex];
  }

  function _getPointerStatus(objectIndex) {
    const pointer = valueTable[objectIndex];
    const status = pointer[ZIG];
    if (typeof(status) !== 'boolean') {
      return -1;
    }
    return status ? 1 : 0;
  }

  function _setPointerStatus(objectIndex, status) {
    const pointer = valueTable[objectIndex];
    pointer[ZIG] = !!status;
  }

  function _readGlobalSlot(slot) {
    const object = globalSlots[slot];
    return object ? valueIndices.get(object) : 0;
  }

  function _writeGlobalSlot(slot, valueIndex) {
    globalSlots[slot] = valueTable[valueIndex];
  }

  function _readObjectSlot(objectIndex, slot) {
    const object = valueTable[objectIndex];
    const value = object[SLOTS][slot];
    return value ? valueIndices.get(value) : 0;
  }

  function _writeObjectSlot(objectIndex, slot, valueIndex) {
    const object = valueTable[objectIndex];
    object[SLOTS][slot] = valueTable[valueIndex];
  }

  function _beginStructure(defIndex) {
    const def = valueTable[defIndex];
    return addObject(beginStructure(def));
  }

  function _attachMember(structureIndex, defIndex) {
    const structure = valueTable[structureIndex];
    const def = valueTable[defIndex];
    attachMember(structure, def);
  }

  function _attachMethod(structureIndex, defIndex) {
    const structure = valueTable[structureIndex];
    const def = valueTable[defIndex];
    def.thunk = { index: nextMethodIndex++ };
    attachMethod(structure, def);
  }

  function _attachTemplate(structureIndex, templateIndex) {
    const structure = valueTable[structureIndex];
    const template = valueTable[templateIndex];
    attachTemplate(structure, template);
  }

  function _finalizeStructure(structureIndex) {
    const structure = valueTable[structureIndex];
    structures.push(structure);
  }

  function _createDataView(ctxAddr, address, len, onStack) {
    // look for address among existing buffers
    const ctx = callContexts[ctxAddr];
    let dv;
    for (const [ buffer, { address: start, len: count } ] of ctx.bufferMap) {
      if (start <= address && address + len <= start + count) {
        const offset = address - start;
        dv = new DataView(buffer, offset, len);
      }
    }
    if (!dv) {
      if (onStack || process.env.NODE_ZIG_TARGET === 'WASM-STAGE1') {
        const buffer = new ArrayBuffer(len);
        const copy = getMemoryCopier(len);
        dv = new DataView(buffer);
        ctx.bufferMap.set(buffer, { address, len, dv, copy });
        if (!onStack) {
          // flag the need for linkage
          dv.linkage = nextLinkageIndex++;
        }
      } else {
        // mystery memory--link directly to it, attaching the memory object
        // so we can recreate the view in the event of buffer deattachment
        // due to address space enlargement
        dv = new DataView(wasmMemory.buffer, address, len);
        dv[SOURCE] = wasmMemory;
      }
    }
    return addObject(dv);
  }

  function _createTemplate(memoryIndex) {
    const memory = valueTable[memoryIndex];
    return addObject({
      [MEMORY]: memory,
      [SLOTS]: {},
    });
  }

  function _createArray() {
    return addObject([]);
  }

  function _appendArray(arrayIndex, valueIndex) {
    const array = valueTable[arrayIndex];
    const value = valueTable[valueIndex];
    return array.push(value);
  }

  function _logValues(arrayIndex) {
    const array = valueTable[arrayIndex];
    console.log(...array);
  }
}

export function serializeDefinitions(structures, params) {
  const {
    runtimeURL,
    loadWASM,
  } = params;
  const lines = [];
  let indent = 0;
  function add(s) {
    if (/^\s*[\]\}]/.test(s)) {
      indent--;
    }
    lines.push(' '.repeat(indent * 2) + s);
    if (/[\[\{]\s*$/.test(s)) {
      indent++;
    }
  }
  const structureFeatures = {}, memberFeatures = {};
  for (const structure of structures) {
    structureFeatures[getStructureFeature(structure)] = true;
    for (const members of [ structure.instance.members, structure.static.members ]) {
      for (const member of members) {
        memberFeatures[ getMemberFeature(member) ] = true;
      }
    }
  }
  if (memberFeatures.useIntEx) {
    delete memberFeatures.useInt;
  }
  if (memberFeatures.useEnumerationItemEx) {
    delete memberFeatures.useEnumerationItem;
  }
  if (memberFeatures.useFloatEx) {
    delete memberFeatures.useFloat;
  }
  if (memberFeatures.useBoolEx) {
    delete memberFeatures.useBool;
  }
  const features = [ ...Object.keys(structureFeatures), ...Object.keys(memberFeatures) ];
  const imports = [ 'finalizeStructures' ];
  if (loadWASM) {
    imports.push('linkWASMBinary');
  }
  imports.push(...features);
  add(`import {`);
  for (const name of imports) {
    add(`${name},`);
  }
  add(`} from ${JSON.stringify(runtimeURL)};`);

  add(`\n// activate features`);
  for (const feature of features) {
    add(`${feature}();`);
  }

  add(`\n// define structures`);
  const structureNames = new Map();
  for (const [ index, structure ] of structures.entries()) {
    const varname = `s${index}`;
    addStructure(varname, structure);
    structureNames.set(structure, varname);
  }

  add(`\n// finalize structures`);
  const varnames = [ ...structureNames.values() ];
  if (varnames.length <= 10) {
    add(`const structures = [ ${varnames.join(', ') } ];`);
  } else {
    add(`const structures = [`);
    for (let i = 0; i < varnames.length; i += 10) {
      const slice = varnames.slice(i, i + 10);
      add(`${slice.join(', ')},`);
    }
    add(`];`);
  }
  add(`const linkage = finalizeStructures(structures);`);

  // the root structure gets finalized last
  const root = structures[structures.length - 1];
  add(`const module = ${structureNames.get(root)}.constructor;`);

  if (loadWASM) {
    add('\n// initialize loading and compilation of WASM bytecodes');
    add(`const binaryPromise = ${loadWASM};`);
    // TODO: figure out how best to load the binary
    add('const __init = linkWASMBinary(binaryPromise, linkage);');
  } else {
    add('\n// no need to initialize WASM binary');
    add('const __init = Promise.resolve(true);');
  }

  add('\n// export functions, types, and constants');
  const exportables = [];
  for (const method of root.methods) {
    exportables.push(method.name);
  }
  for (const member of root.static.members) {
    // only read-only properties are exportable
    let readOnly = false;
    if (member.type === MemberType.Type) {
      readOnly = true;
    } else if (member.type === MemberType.Object && member.structure.type === StructureType.Pointer) {
      if (member.isConst) {
        readOnly = true;
      }
    }
    if (readOnly) {
      exportables.push(member.name);
    }
  }
  add(`const {`);
  for (const name of exportables) {
    add(`${name},`);
  }
  add(`} = module;`);
  add(`export {`);
  for (const name of [ 'module as default', ...exportables, '__init' ]) {
    add(`${name},`);
  }
  add(`};`);
  add(``);

  function addStructure(varname, structure) {
    add(`const ${varname} = {`);
    for (const [ name, value ] of Object.entries(structure)) {
      switch (name) {
        case 'instance':
        case 'static':
          add(`${name}: {`);
          addMembers(value.members);
          addTemplate(value.template);
          add(`},`);
          break;
        case 'methods':
          addMethods(value);
          break;
        default:
          add(`${name}: ${JSON.stringify(value)},`);
      }
    }
    add(`};`);
  }

  function addMembers(members) {
    if (members.length > 0) {
      add(`members: [`);
      for (const member of members) {
        addMember(member);
      }
      add(`],`);
    } else {
      add(`members: [],`);
    }
  }

  function addMember(member) {
    add(`{`);
    for (const [ name, value ] of Object.entries(member)) {
      switch (name) {
        case 'structure':
          add(`${name}: ${structureNames.get(value)},`);
          break;
        default:
          add(`${name}: ${JSON.stringify(value)},`);
      }
    }
    add(`},`);
  }

  function addTemplate(template) {
    addObject('template', template);
  }

  function addObject(name, object) {
    if (object) {
      const { [STRUCTURE]: structure, [MEMORY]: dv, [SLOTS]: slots } = object;
      add(`${name}: {`);
      if (structure) {
        add(`structure: ${structureNames.get(structure)},`);
      }
      if (dv) {
        const ta = new Uint8Array(dv.buffer, dv.byteOffset, dv.byteLength);
        add(`memory: [ ${ta.join(', ')} ],`);
        if (dv.hasOwnProperty('linkage')) {
          add(`linkage: ${dv.linkage},`);
        }
      }
      if (slots && Object.keys(slots).length > 0) {
        add(`slots: {`);
        for (const [ slot, object ] of Object.entries(slots)) {
          addObject(slot, object);
        }
        add(`},`);
      }
      add(`},`);
    } else {
      add(`${name}: null`);
    }
  }

  function addMethods(methods) {
    if (methods.length > 0) {
      add(`methods: [`);
      for (const method of methods) {
        addMethod(method);
      }
      add(`],`);
    } else {
      add(`methods: [],`)
    }
  }

  function addMethod(method) {
    add(`{`);
    for (const [ name, value ] of Object.entries(method)) {
      switch (name) {
        case 'argStruct':
          add(`${name}: ${structureNames.get(value)},`);
          break;
        default:
          add(`${name}: ${JSON.stringify(value)},`);
      }
    }
    add(`},`);
  }

  return lines.join('\n');
}

export function finalizeStructures(structures) {
  const variables = [];
  for (const structure of structures) {
    for (const target of [ structure.static, structure.instance ]) {
      // first create the actual template using the provided placeholder
      if (target.template) {
        target.template = createTemplate(target.template);
      }
    }
    for (const method of structure.methods) {
      // create thunk function
      method.thunk = createThunk(method.thunk);
    }
    finalizeStructure(structure);
  }

  function createTemplate(placeholder) {
    const template = {};
    if (placeholder.memory) {
      template[MEMORY] = new DataView(new Uint8Array(placeholder.memory).buffer);
    }
    if (placeholder.slots) {
      template[SLOTS] = insertObjects({}, placeholder.slots);
    }
    if (placeholder.linkage !== undefined) {
      // need to replace dataview with one pointing to WASM memory later,
      // when the VM is up and running
      variables[placeholder.linkage] = template;
    }
    return template;
  }

  function insertObjects(dest, placeholders) {
    for (const [ slot, placeholder ] of Object.entries(placeholders)) {
      dest[slot] = createObject(placeholder);
    }
    return dest;
  }

  function createObject(placeholder) {
    const dv = new DataView(new Uint8Array(placeholder.memory).buffer);
    const { constructor } = placeholder.structure;
    const object = constructor.call(null, dv);
    if (placeholder.slots) {
      insertObjects(object[SLOTS], placeholder.slots);
    }
    if (placeholder.linkage !== undefined) {
      variables[placeholder.linkage] = object;
    }
    return object;
  }

  let resolve, reject;
  const promise = new Promise((r1, r2) => {
    resolve = r1;
    reject = r2;
  });
  const methodRunner = {
    0: function(index, argStruct) {
      // wait for linking to occur, then active the runner again
      return promise.then(() => methodRunner[0].call(this, index, argStruct));
    },
  };

  function createThunk({ index }) {
    return function(argStruct) {
      return methodRunner[0](index, argStruct);
    };
  }

  return { promise, resolve, reject, variables, methodRunner };
}

export {
  usePrimitive,
  useArray,
  useStruct,
  useExternUnion,
  useBareUnion,
  useTaggedUnion,
  useErrorUnion,
  useErrorSet,
  useEnumeration,
  useOptional,
  usePointer,
  useSlice,
  useOpaque,
  useArgStruct,
} from './structure.js';
export {
  useVoid,
  useBool,
  useBoolEx,
  useInt,
  useIntEx,
  useFloat,
  useFloatEx,
  useEnumerationItem,
  useEnumerationItemEx,
  useObject,
  useType,
} from './member.js';

File: src/wasm-exporter.zig
const std = @import("std");
const exporter = @import("exporter.zig");

const Result = exporter.Result;
const Host = exporter.Host;
const Value = exporter.Value;
const Module = exporter.Module;
const StructureType = exporter.StructureType;
const Structure = exporter.Structure;
const MemberType = exporter.MemberType;
const Member = exporter.Member;
const Method = exporter.Method;
const Memory = exporter.Memory;
const MemoryDisposition = exporter.MemoryDisposition;
const Template = exporter.Template;
const Thunk = exporter.Thunk;
const missing = exporter.missing;

const CallContext = struct {
    allocator: std.mem.Allocator,
};

fn ref(number: usize) Value {
    return @ptrFromInt(number);
}

fn index(object: ?Value) usize {
    return if (object) |ptr| @intFromPtr(ptr) else 0;
}

fn strlen(s: [*:0]const u8) usize {
    var len: usize = 0;
    while (s[len] != 0) {
        len += 1;
    }
    return len;
}

extern fn _createObject() usize;

fn createObject() Value {
    return ref(_createObject());
}

extern fn _createString(address: usize, len: usize) usize;

extern fn _setObjectPropertyString(container: usize, key: usize, value: usize) void;
extern fn _setObjectPropertyInteger(container: usize, key: usize, value: i32) void;
extern fn _setObjectPropertyBoolean(container: usize, key: usize, value: i32) void;
extern fn _setObjectPropertyObject(container: usize, key: usize, value: usize) void;

fn setObjectProperty(container: Value, key: []const u8, value: anytype) void {
    const T = @TypeOf(value);
    const key_index = _createString(@intFromPtr(key.ptr), key.len);
    switch (@typeInfo(T)) {
        .Pointer => {
            if (T == [*:0]const u8) {
                const value_index = _createString(@intFromPtr(value), strlen(value));
                _setObjectPropertyString(index(container), key_index, value_index);
            } else if (T == Value) {
                _setObjectPropertyObject(index(container), key_index, index(value));
            } else {
                @compileError("No support for value type: " ++ @typeName(T));
            }
        },
        .Int => {
            _setObjectPropertyInteger(index(container), key_index, @intCast(value));
        },
        .Enum => {
            _setObjectPropertyInteger(index(container), key_index, @intCast(@intFromEnum(value)));
        },
        .Bool => {
            _setObjectPropertyBoolean(index(container), key_index, if (value) 1 else 0);
        },
        .Optional => {
            if (value) |v| {
                setObjectProperty(container, key, v);
            } else {
                _setObjectPropertyObject(index(container), key_index, 0);
            }
        },
        else => {
            @compileError("No support for value type: " ++ @typeName(T));
        },
    }
}

export fn alloc(host: Host, len: usize) usize {
    const ctx: *CallContext = @ptrCast(@alignCast(@constCast(host)));
    if (len > 0) {
        if (ctx.allocator.alloc([]u8, len)) |bytes| {
            return @intFromPtr(bytes.ptr);
        } else |_| {
            return 0;
        }
    } else {
        return 0;
    }
}

export fn free(host: Host, address: usize, len: usize) void {
    const ctx: *CallContext = @ptrCast(@alignCast(@constCast(host)));
    const bytes: [*]u8 = @ptrFromInt(address);
    ctx.allocator.free(bytes[0..len]);
}

extern fn _allocMemory(host: usize, size: usize) usize;

fn allocateMemory(host: Host, size: usize, memory: *Memory) callconv(.C) Result {
    const address = _allocMemory(@intFromPtr(host), size);
    if (address == 0) {
        return Result.Failure;
    }
    memory.* = .{
        .bytes = @ptrFromInt(address),
        .len = size,
    };
    return Result.OK;
}

extern fn _freeMemory(host: usize, address: usize, len: usize) void;

fn freeMemory(host: Host, memory: *const Memory) callconv(.C) Result {
    _freeMemory(@intFromPtr(host), @intFromPtr(memory.bytes), memory.len);
    return Result.OK;
}

extern fn _getMemory(host: usize, object: usize) usize;
extern fn _getMemoryOffset(object: usize) usize;
extern fn _getMemoryLength(object: usize) usize;

fn getMemory(host: Host, object: Value, memory: *Memory) callconv(.C) Result {
    const view_index = _getMemory(@intFromPtr(host), index(object));
    if (view_index == 0) {
        return Result.Failure;
    }
    const len = _getMemoryLength(view_index);
    const offset = _getMemoryOffset(view_index);
    memory.* = .{
        .bytes = @ptrFromInt(offset),
        .len = len,
    };
    return Result.OK;
}

extern fn _createDataView(host: usize, address: usize, len: usize, disposition: u32) usize;

fn createDataView(host: Host, memory: *const Memory, disposition: MemoryDisposition) ?Value {
    const bytes = memory.bytes orelse return null;
    const len = memory.len;
    const address = @intFromPtr(bytes);
    const dv_index = _createDataView(@intFromPtr(host), address, len, @intFromEnum(disposition));
    if (dv_index == 0) {
        return null;
    }
    return ref(dv_index);
}

extern fn _wrapMemory(structure: usize, view: usize) usize;

fn wrapMemory(host: Host, structure: Value, memory: *const Memory, disposition: MemoryDisposition, dest: *Value) callconv(.C) Result {
    const dv = createDataView(host, memory, disposition) orelse return Result.Failure;
    const obj_index = _wrapMemory(index(structure), index(dv));
    if (obj_index == 0) {
        return Result.Failure;
    }
    dest.* = ref(obj_index);
    return Result.OK;
}

extern fn _getPointerStatus(object: usize) i32;

fn getPointerStatus(_: Host, object: Value, status: *bool) callconv(.C) Result {
    const value = _getPointerStatus(index(object));
    if (value == -1) {
        return Result.Failure;
    }
    status.* = (value != 0);
    return Result.OK;
}

extern fn _setPointerStatus(object: usize, status: i32) void;

fn setPointerStatus(_: Host, object: Value, status: bool) callconv(.C) Result {
    _setPointerStatus(index(object), if (status) 1 else 0);
    return Result.OK;
}

extern fn _readGlobalSlot(slot: usize) usize;

fn readGlobalSlot(_: Host, slot: usize, dest: *Value) callconv(.C) Result {
    const obj_index = _readGlobalSlot(slot);
    if (obj_index == 0) {
        return Result.Failure;
    }
    dest.* = ref(obj_index);
    return Result.OK;
}

extern fn _writeGlobalSlot(slot: usize, object: usize) void;

fn writeGlobalSlot(_: Host, slot: usize, object: ?Value) callconv(.C) Result {
    _writeGlobalSlot(@truncate(slot), index(object));
    return Result.OK;
}

extern fn _readObjectSlot(container: usize, slot: usize) usize;

fn readObjectSlot(_: Host, container: Value, slot: usize, dest: *Value) callconv(.C) Result {
    const obj_index = _readObjectSlot(index(container), slot);
    if (obj_index == 0) {
        return Result.Failure;
    }
    dest.* = ref(obj_index);
    return Result.OK;
}

extern fn _writeObjectSlot(container: usize, slot: usize, object: usize) void;

fn writeObjectSlot(_: Host, container: Value, slot: usize, object: ?Value) callconv(.C) Result {
    _writeObjectSlot(index(container), slot, index(object));
    return Result.OK;
}

extern fn _beginStructure(def: u32) u32;

fn beginStructure(_: Host, structure: *const Structure, dest: *Value) callconv(.C) Result {
    const def = createObject();
    setObjectProperty(def, "name", structure.name);
    setObjectProperty(def, "type", structure.structure_type);
    setObjectProperty(def, "size", structure.total_size);
    setObjectProperty(def, "hasPointer", structure.has_pointer);
    dest.* = ref(_beginStructure(index(def)));
    return Result.OK;
}

extern fn _attachMember(structure: usize, def: usize) void;

fn attachMember(_: Host, structure: Value, member: *const Member) callconv(.C) Result {
    const def = createObject();
    setObjectProperty(def, "type", member.member_type);
    setObjectProperty(def, "isSigned", member.is_signed);
    setObjectProperty(def, "isConst", member.is_const);
    setObjectProperty(def, "isRequired", member.is_required);
    setObjectProperty(def, "isStatic", member.is_static);
    if (member.bit_offset != missing) {
        setObjectProperty(def, "bitOffset", member.bit_offset);
    }
    if (member.bit_size != missing) {
        setObjectProperty(def, "bitSize", member.bit_size);
    }
    if (member.byte_size != missing) {
        setObjectProperty(def, "byteSize", member.byte_size);
    }
    if (member.slot != missing) {
        setObjectProperty(def, "slot", member.slot);
    }
    if (member.name) |name| {
        setObjectProperty(def, "name", name);
    }
    if (member.structure) |s| {
        setObjectProperty(def, "structure", s);
    }
    _attachMember(index(structure), index(def));
    return Result.OK;
}

extern fn _attachMethod(structure: usize, def: usize) void;

fn attachMethod(_: Host, structure: Value, method: *const Method) callconv(.C) Result {
    const def = createObject();
    setObjectProperty(def, "isStatic", method.is_static_only);
    setObjectProperty(def, "argStruct", method.structure);
    setObjectProperty(def, "thunk", @intFromPtr(method.thunk));
    if (method.name) |name| {
        setObjectProperty(def, "name", name);
    }
    _attachMethod(index(structure), index(def));
    return Result.OK;
}

extern fn _attachTemplate(structure: usize, def: usize) void;

fn attachTemplate(_: Host, structure: Value, template: *const Template) callconv(.C) Result {
    const def = createObject();
    setObjectProperty(def, "isStatic", template.is_static);
    setObjectProperty(def, "template", template.object);
    _attachTemplate(index(structure), index(def));
    return Result.OK;
}

extern fn _finalizeStructure(structure: usize) void;

fn finalizeStructure(_: Host, structure: Value) callconv(.C) Result {
    _finalizeStructure(index(structure));
    return Result.OK;
}

extern fn _createTemplate(buffer: usize) usize;

fn createTemplate(host: Host, memory: *const Memory, dest: *Value) callconv(.C) Result {
    const dv = createDataView(host, memory, .Copy);
    const templ_index = _createTemplate(index(dv));
    if (templ_index == 0) {
        return Result.Failure;
    }
    dest.* = ref(templ_index);
    return Result.OK;
}

fn createString(_: Host, memory: *const Memory, dest: *Value) callconv(.C) Result {
    const address = @intFromPtr(memory.bytes);
    const len = memory.len;
    dest.* = ref(_createString(address, len));
    return Result.OK;
}

extern fn _createArray() usize;
extern fn _appendArray(array: usize, value: usize) void;
extern fn _logValues(values: usize) void;

fn logValues(_: Host, argc: usize, argv: [*]Value) callconv(.C) Result {
    const array = _createArray();
    var i: usize = 0;
    while (i < argc) : (i += 1) {
        _appendArray(array, index(argv[i]));
    }
    _logValues(array);
    return Result.OK;
}

pub fn setCallbacks() void {
    const ptr = &exporter.callbacks;
    ptr.allocate_memory = allocateMemory;
    ptr.free_memory = freeMemory;
    ptr.get_memory = getMemory;
    ptr.wrap_memory = wrapMemory;
    ptr.get_pointer_status = getPointerStatus;
    ptr.set_pointer_status = setPointerStatus;
    ptr.read_global_slot = readGlobalSlot;
    ptr.write_global_slot = writeGlobalSlot;
    ptr.read_object_slot = readObjectSlot;
    ptr.write_object_slot = writeObjectSlot;
    ptr.begin_structure = beginStructure;
    ptr.attach_member = attachMember;
    ptr.attach_method = attachMethod;
    ptr.attach_template = attachTemplate;
    ptr.finalize_structure = finalizeStructure;
    ptr.create_template = createTemplate;
    ptr.create_string = createString;
    ptr.log_values = logValues;
}

extern fn _startCall(host: usize) void;
extern fn _endCall(host: usize) void;

pub fn runThunk(arg_index: usize, address: usize) usize {
    var ctx: CallContext = .{
        .allocator = .{ .ptr = undefined, .vtable = &std.heap.WasmAllocator.vtable },
    };
    const thunk: Thunk = @ptrFromInt(address);
    const host: Host = @ptrCast(&ctx);
    _startCall(@intFromPtr(host));
    const result = thunk(host, ref(arg_index));
    _endCall(@intFromPtr(host));
    if (result) |error_msg| {
        return _createString(@intFromPtr(error_msg), strlen(error_msg));
    } else {
        return 0;
    }
}

pub fn exportModule(comptime T: type, arg_index: usize) usize {
    const factory = exporter.createRootFactory(T);
    return runThunk(arg_index, @intFromPtr(factory));
}
